### ISSUE TITLE
Unix.read exhibits two different behaviors on Windows vs Linux

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#7342]
**Original bug ID:** 7342
**Reporter:** @protz
**Assigned to:** @protz
**Status:** resolved (set by @gasche on 2016-12-16T02:49:50Z)
**Resolution:** fixed
**Priority:** normal
**Severity:** minor
**Version:** 4.05.0 +dev/beta1/beta2/beta3/rc1
**Target version:** 4.05.0 +dev/beta1/beta2/beta3/rc1
**Fixed in version:** 4.05.0 +dev/beta1/beta2/beta3/rc1
**Category:** platform support (windows, cross-compilation, etc)
**Monitored by:** dsheets @gasche @hcarty

## Bug description

Reading Didier and Xavier's excellent "Unix programming in OCaml" course, I performed a `Unix.read` on a descriptor, expecting to read 0 when the descriptor had nothing left in it.

Sadly, I was developing on Windows. It turns out that `Unix.read` throws `EPIPE` there instead (see https://github.com/ocsigen/lwt/issues/20 for another project that performed a workaround there).

## Steps to reproduce

See the attached file.
```ocaml
# #use "Test.ml";;
- : unit = ()

# run "ls" [||];;
Exception: Unix.Unix_error (Unix.EPIPE, "read", "").
```
The workaround is to remove the "when false" clause in Test.ml.

## Additional information

I can submit a pull request that wraps otherlibs/win32unix/unix.ml:read with that exact try-catch block, or I can submit a documentation request that documents this caveat. Any preference?

## File attachments

- [Test.ml](https://gist.githubusercontent.com/vicuna/c789bb3ba0f381ac7ab9a5f1763b16db/raw/67f185b2f26c042289ade9675f3bb0509362ce44/Test.ml)

**Comment by vicuna** [comment]
**Comment author:** @gasche

As a mostly-unix programmer I would have a preference for the documented behavior (also in `man 3 read` on the C side) to work reliably on Windows -- I'm sure I would have been confused by the issue if people tried my software on Windows.

**Comment by vicuna** [comment]
**Comment author:** @xavierleroy

Another Win32/Unix gratuitous difference, wonderful.

My preferred fix would be to change win32unix/read.c so that a GetLastError() code ERROR_BROKEN_PIPE is turned into a return of 0.

**Comment by vicuna** [comment]
**Comment author:** @protz

https://github.com/ocaml/ocaml/pull/797

**Comment by vicuna** [comment]
**Comment author:** @gasche

Fixed by merging #797.
